Update README (#80)
simplabs/ember-promise-modals: README.md

ember-promise-modals
==============================================================================

> Ember promise modal (EPM) provide a way to render and handle multiple modals via Javascript.

Reviewer: ```suggestion
ember-promise-modals provides a way to render and handle modals in Ember.js apps.
```

* Ember CLI v2.13 or above
* Node.js v8 or above

Installation
------------------------------------------------------------------------------

```
ember install ember-promise-modals
```

Usage
------------------------------------------------------------------------------

To use EPM in your project, you first need to inject the `modals` service:
```javascript
modals: inject()
```

> Then, call the `open` method with a component name to render in the modal:

Reviewer: ```suggestion
Then, call the `open` method with a component name to render it as a modal:
```

this.modals.open('component-to-render');
```

### Simple usage

Here is a simple code example:

```javascript
import { inject } from '@ember/service';
...
modals: inject(),
...
actions: {
    handleOpenModal() {
        this.modals.open('confirmation-modal');
    }
}
```

Then in your template, you can:

```html
<button {{on "click" this.handleOpenModal}}>
    Click Me!
</button>
```

### Attributes

You can pass custom data into your rendered template like so:
```javascript
this.modals.open('file-preview', {
    fileUrl: this.fileUrl
});
```

All passed attributes can be accessed from the `data` object:

```html
<!-- app/components/file-preview.hbs -->

<img src={{@data.fileUrl}} />
```

```javascript
// app/components/file-preview.js

this.data.fileUrl;
```

**NOTE:** By default, a `close` method is passed in your rendered component, in order to trigger the "close modal" action. It can be called like so:

```html
<!-- app/components/file-preview.hbs -->

<button {{on "click" @close}}>Close</button>
```
```javascript
// app/components/file-preview.js

this.close();
```

Animation
------------------------------------------------------------------------------

### Custom animation

> EPM uses [ember-animated](https://github.com/ember-animation/ember-animated) internally in order to animate modals transition. You can also use animations from libraries like [d3-ease](https://github.com/d3/d3-ease)

Reviewer: `d3-ease` provides easing functions, not animations directly

Author: Sure, will update

Author: Updated

Author: Updated https://github.com/simplabs/ember-promise-modals/pull/80/files#diff-04c6e90faac2675aa89e2176d2eec7d8R100-R102

```javascript
this.modals.open('file-preview', {
    reverseAnimation: true
});
```

And update your service:

```javascript
insertedSprites.forEach(sprite => {
    if (sprite.owner.value._data.reverseAnimation) {
        sprite.startAtPixel({ y: window.innerHeight });
    } else {
        sprite.startAtPixel({ y: -window.innerHeight });
    }
    move(sprite, { easing: easeCubicOut });
});

...

removedSprites.forEach(sprite => {
    if (sprite.owner.value._data.reverseAnimation) {
        sprite.endAtPixel({ y: window.innerHeight });
    } else {
        sprite.endAtPixel({ y: -window.innerHeight });
    }
    move(sprite, { easing: easeSinInOut });
});
```

You can find other way to create custom animations in [ember-animated documentation](https://ember-animation.github.io/ember-animated/docs/transitions#custom)

Accessibility
------------------------------------------------------------------------------

User can press `ESC` key to close the modal.

EPM uses [focus-trap](https://github.com/davidtheclark/focus-trap) internally in order to handle user focus.

> User will [focus the first "tabbable element" by default](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). If no focusable element is present, focus will be applied on the currently visible modal amber-auto-generated container.

Reviewer: what do you mean by "amber-auto-generated container"?

Author: Not really explicit indeed ! I was talking about the auto generated div with a random `ember-xxx` id that wrap any component without `tagName: ''`

Author: What do you think of 
```suggestion
User will [focus the first "tabbable element" by default](https://www.w3.org/TR/wai-aria-practices-1.1/#dialog_modal). If no focusable element is present, focus will be applied on the modal container, auto-generated by Ember.
```
